src/database_migration.py

```diff
@@ -0,0 +1,142 @@
+
+logger = logging.getLogger(__name__)
+
+
+DatabaseMigrationStatus = Literal["PENDING", "COMPLETED", "FAILED"]
+
+
+class DatabaseMigration:
+    """The DatabaseMigration class that manages database migrations.
+
+    Attrs:
+        script: the database migration script.
+        COMPLETED: one of the possible database migration status.
+        FAILED: one of the possible database migration status.
+        PENDING: one of the possible database migration status.
+    """
+
+    _STATUS_FILE = FLASK_STATE_DIR / "database-migration-status"
+    _COMPLETED_SCRIPT_FILE = FLASK_STATE_DIR / "completed-database-migration"
+
+    COMPLETED: DatabaseMigrationStatus = "COMPLETED"
+    FAILED: DatabaseMigrationStatus = "FAILED"
+    PENDING: DatabaseMigrationStatus = "PENDING"
+
+    def __init__(self, flask_container: ops.Container, database_migration_script: str | None):
+        """Initialize the DatabaseMigration instance.
+
+        Args:
+            flask_container: The flask container object.
+            database_migration_script: The database_migration_script configuration.
+        """
+        self._container = flask_container
+        self._script = database_migration_script
+
+    @property
+    def script(self) -> str | None:
+        """Get the database migration script."""
+        return self._script
+
+    def get_status(self) -> DatabaseMigrationStatus:
+        """Get the database migration run status.
+
+        Returns:
+            One of "PENDING", "COMPLETED", or "FAILED".
+        """
+        return cast(
+            DatabaseMigrationStatus,
+            self.PENDING
+            if not self._container.exists(self._STATUS_FILE)
+            else self._container.pull(self._STATUS_FILE).read(),
+        )
+
+    def set_status(self, status: DatabaseMigrationStatus) -> None:
+        """Set the database migration run status.
+
+        Args:
+            status: One of "PENDING", "COMPLETED", or "FAILED".
+        """
+        self._container.make_dir(self._STATUS_FILE.parent, make_parents=True)
+        self._container.push(self._STATUS_FILE, source=status)
+
+    def get_completed_script(self) -> str | None:
+        """Get the database migration script that has completed in the current container.
+
+        Returns:
+            The completed database migration script in the current container.
+        """
+        if self._container.exists(self._COMPLETED_SCRIPT_FILE):
+            return cast(str, self._container.pull(self._COMPLETED_SCRIPT_FILE).read())
+        return None
+
+    def set_completed_script(self, script_path: str) -> None:
+        """Set the database migration script that has completed in the current container.
+
+        Args:
+            script_path: The completed database migration script in the current container.
+        """
+        self._container.make_dir(self._COMPLETED_SCRIPT_FILE.parent, make_parents=True)
+        self._container.push(self._COMPLETED_SCRIPT_FILE, script_path)
```

Comment:
What is the reason for needing these please?

Reply:
perhaps they can be private?

Reply:
I was more talking about whether `set_completed_script` and `get_completed_script` could be private methods?

Reply:
See https://github.com/canonical/flask-k8s-operator/pull/50#discussion_r1320814651, thanks!